### Lab 3.1: Basic Neural Network in PyTorch

### Exercises

1. Extend the code from Notebook 3.1 to implement an MLP with a single hidden layer of size 100.

In [1]:
pip install palmerpenguins

In [6]:
import numpy as np
import torch
from palmerpenguins import load_penguins
from matplotlib import pyplot as plt

df = load_penguins()

# drop rows with missing values
df.dropna(inplace=True)

# get two features
X = df[['flipper_length_mm','bill_length_mm']].values

# convert species labels to integers
y = df['species'].map({'Adelie':0,'Chinstrap':1,'Gentoo':2}).values

X -= np.mean(X,axis=0)

X = torch.tensor(X).float()
y = torch.tensor(y).long()

MLP_model = torch.nn.Sequential(
    torch.nn.Linear(2,100), # two inputs, 100 outputs
    torch.nn.ReLU(), # activation function
    torch.nn.Linear(100,3) # 100 inputs, 3 outputs
)

linear_model = torch.nn.Sequential(
    torch.nn.Linear(2,3) # two inputs, 3 outputs
)

loss_fn = torch.nn.CrossEntropyLoss()

lr = 1e-2
MLP_opt = torch.optim.SGD(MLP_model.parameters(), lr=lr)
linear_opt = torch.optim.SGD(linear_model.parameters(), lr=lr)

epochs = 100

print("-----------LINEAR MODEL---------------")
for epoch in range(epochs):
    linear_opt.zero_grad() # zero out the gradients

    z = linear_model(X) # compute z values
    loss = loss_fn(z,y) # compute loss

    loss.backward() # compute gradients

    linear_opt.step() # apply gradients

    print(f'epoch {epoch}: loss is {loss.item()}')

print("-----------MLP MODEL---------------")
for epoch in range(epochs):
    MLP_opt.zero_grad() # zero out the gradients

    z = MLP_model(X) # compute z values
    loss = loss_fn(z,y) # compute loss

    loss.backward() # compute gradients

    MLP_opt.step() # apply gradients

    print(f'epoch {epoch}: loss is {loss.item()}')

-----------LINEAR MODEL---------------
epoch 0: loss is 2.7544989585876465
epoch 1: loss is 2.461244821548462
epoch 2: loss is 2.24544095993042
epoch 3: loss is 2.0851211547851562
epoch 4: loss is 1.9619985818862915
epoch 5: loss is 1.863647222518921
epoch 6: loss is 1.7816945314407349
epoch 7: loss is 1.7106297016143799
epoch 8: loss is 1.6469347476959229
epoch 9: loss is 1.5883818864822388
epoch 10: loss is 1.5335452556610107
epoch 11: loss is 1.4814943075180054
epoch 12: loss is 1.4316096305847168
epoch 13: loss is 1.3834710121154785
epoch 14: loss is 1.3367886543273926
epoch 15: loss is 1.2913621664047241
epoch 16: loss is 1.2470533847808838
epoch 17: loss is 1.2037684917449951
epoch 18: loss is 1.1614465713500977
epoch 19: loss is 1.120052456855774
epoch 20: loss is 1.079569697380066
epoch 21: loss is 1.0399976968765259
epoch 22: loss is 1.0013477802276611
epoch 23: loss is 0.9636420011520386
epoch 24: loss is 0.9269100427627563
epoch 25: loss is 0.8911880850791931
epoch 26: loss 

2. Write code to compute the accuracy of each model.


In [7]:
def accuracy(model, X, y):
    model.eval()
    with torch.no_grad():
        z = model(X)
        preds = z.argmax(dim=1)
        return (preds == y).float().mean().item()

print(f'linear model accuracy: {accuracy(linear_model, X, y):.4f}')

linear model accuracy: 0.9429


In [8]:
def accuracy(model, X, y):
    model.eval()
    with torch.no_grad():
        z = model(X)
        preds = z.argmax(dim=1)
        return (preds == y).float().mean().item()

print(f'MLP accuracy: {accuracy(MLP_model, X, y):.4f}')

MLP accuracy: 0.9520


3. Compare the MLP and linear models in terms of accuracy on the training set.

The MLP got a training accuracy of 0.9520, and the linear model got a training accuracy of 0.9249 for the linear model. This makes sense for training accuracy as, due to the added complexity that the MLP has, it separated the data better, not being restricted to only linear boundaries. We don't know off of the training accuracy which model generalizes to new data better.